### PR TITLE
Ensure selection of instances does not contain deleted global objects

### PR DESCRIPTION
**Pull request by 4ian**
- Also keep the selection of objects/folders (unless they have been deleted)

### CODE REVIEW

File: newIDE/app/src/InstancesEditor/InstancesSelection.spec.js [hunk 1]
```diff
@@ -353,4 +353,30 @@ describe('InstancesSelection', () => {
     instance2OfObject2.delete();
     instance3OfObject2.delete();
   });
+
+  it('can clean non existing instances', () => {
```
Comment (AlexandreSi) [inline newIDE/app/src/InstancesEditor/InstancesSelection.spec.js:357]: Clean tests

File: newIDE/app/src/SceneEditor/index.js [hunk 2]
```diff
@@ -340,7 +341,16 @@ export default class SceneEditor extends React.Component<Props, State> {
       // (global objects in a different scene for instance)
       // - and some components try to access said objects properties
       // (computing selectedObjectNames for instance)
-      this.setState({ selectedObjectFolderOrObjectsWithContext: [] });
```
Comment (AlexandreSi) [inline newIDE/app/src/SceneEditor/index.js:343]: You can change the comment above

**Review by AlexandreSi — APPROVED** [review]
LGTM

File: newIDE/app/src/InstancesEditor/InstancesSelection.js [hunk 3]
```diff
@@ -105,4 +105,25 @@ export default class InstancesSelection {
...
+    functor.delete();
+
+    for (let i = 0; i < this.selection.length; ) {
+      const instance = this.selection[i];
+      if (!allExistingInstancePointers.has(instance.ptr)) {
+        this.selection.splice(i, 1);
+      } else {
+        i++;
+      }
+    }
```
Comment (D8H) [inline newIDE/app/src/InstancesEditor/InstancesSelection.js:127]: What about using a "dead" flag?

**Comment by D8H** [comment]
Strings are often used to refer to instances. For instance a function is "MyExtension::MyBeahvior::MyFunction".
It can't be the same for object instance instances?

**Comment by 4ian** [comment]
Unique ids could work, but at the end the problem is still the same: how do you ensure you keep a selection that does not have outdated, removed stuff?
In case of unique ids, it's less of a problem though, because in the worst case you just try to get the object from the id and it fails and you assume the id is dead. In the case I've fixed, it was worse because the objects looked like they could be manipulated, but were pointed to dead things.

But we'll still go toward an approach that favors unique ids in the future, especially for collaboration stuff. So it's not a bad idea, just that here I wanted to fix the "leak" in the selections :)

**Pull request merged**